File: app/src/main/java/xyz/sleekstats/completist/view/RankingsAdapter.java
package xyz.sleekstats.completist.view;

import android.support.annotation.NonNull;
import android.support.v4.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.woxthebox.draglistview.DragItemAdapter;

import java.util.ArrayList;

import xyz.sleekstats.completist.R;
import xyz.sleekstats.completist.model.MediaByPerson;

class RankingsAdapter extends DragItemAdapter<Pair<Long, MediaByPerson>, RankingsAdapter.ViewHolder> {

    private final int mLayoutId;
    private final int mGrabHandleId;
    private final boolean mDragOnLongPress;
    private final boolean isRanked;
    private ItemClickListener mClickListener;

    RankingsAdapter(ArrayList<Pair<Long, MediaByPerson>> list, boolean ranked) {
        mLayoutId = R.layout.rank_item;
        mGrabHandleId = R.id.grabber;
        mDragOnLongPress = false;
        isRanked = ranked;
        setItemList(list);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(mLayoutId, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        super.onBindViewHolder(holder, position);
        MediaByPerson film = mItemList.get(position).second;
        holder.bindFilm(film);
        if(isRanked) {
            String positionText = String.valueOf(position+1);
            holder.mRankTextView.setText(positionText);
        }
        holder.itemView.setTag(mItemList.get(position));
    }

    @Override
    public long getUniqueItemId(int position) {
        return mItemList.get(position).first;
    }

    public MediaByPerson getFilmOrShow(int position) {
        return mItemList.get(position).second;
    }

    class ViewHolder extends DragItemAdapter.ViewHolder {
        private final TextView mMovieTextView;
        private final TextView mRankTextView;
        private MediaByPerson mFilm;

        ViewHolder(final View itemView) {
            super(itemView, mGrabHandleId, mDragOnLongPress);
            mMovieTextView = itemView.findViewById(R.id.text);
            mRankTextView = itemView.findViewById(R.id.grabber);
        }

        private void bindFilm(MediaByPerson film) {
            mFilm = film;
            mMovieTextView.setText(mFilm.getTitle());
        }

        @Override
        public void onItemClicked(View view) {
        }

        @Override
        public boolean onItemLongClicked(View view) {
            if(mClickListener != null) {
                if(mFilm.isFilm()) {
                    mClickListener.onFilmClick(mFilm.getId());
                } else {
                    mClickListener.onShowClick(mFilm.getId());
                }
            }
            return true;
        }
    }


    public void setClickListener(ItemClickListener itemClickListener) {
        this.mClickListener = itemClickListener;
    }

    public interface ItemClickListener{
        void onFilmClick(String id);
        void onShowClick(String id);
    }

}

File: app/src/main/java/xyz/sleekstats/completist/view/RankingsAdapter.kt
package xyz.sleekstats.completist.view

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.util.Pair
import com.woxthebox.draglistview.DragItemAdapter
import xyz.sleekstats.completist.R
import xyz.sleekstats.completist.model.MediaByPerson
import java.util.*

internal class RankingsAdapter(list: ArrayList<Pair<Long, MediaByPerson>>, ranked: Boolean) : DragItemAdapter<Pair<Long, MediaByPerson>, RankingsAdapter.ViewHolder>() {

    private val mLayoutId: Int = R.layout.rank_item
    private val mGrabHandleId: Int = R.id.grabber
    private val mDragOnLongPress: Boolean = false
    private val isRanked: Boolean = ranked
    private var mClickListener: ItemClickListener? = null

    init {
        itemList = list
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
            ViewHolder(LayoutInflater.from(parent.context).inflate(mLayoutId, parent, false))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)

        holder.bindFilm(mItemList[position]?.second)
        if (isRanked) {
            val positionText = (position + 1).toString()
            holder.mRankTextView.text = positionText
        }
        holder.itemView.tag = mItemList[position]
    }

    override fun getUniqueItemId(position: Int): Long = mItemList[position]?.first!!

    fun getFilmOrShow(position: Int): MediaByPerson = mItemList[position]?.second!!

    internal inner class ViewHolder(itemView: View) : DragItemAdapter.ViewHolder(itemView, mGrabHandleId, mDragOnLongPress) {

        private val mMovieTextView: TextView = itemView.findViewById(R.id.text)
        val mRankTextView: TextView = itemView.findViewById(R.id.grabber)
        private var mFilm: MediaByPerson? = null

        fun bindFilm(film: MediaByPerson?) {
            mFilm = film
            mMovieTextView.text = mFilm?.title
        }

        override fun onItemClicked(view: View) {}

        override fun onItemLongClicked(view: View): Boolean {
            if (mFilm?.isFilm == true) {
                mClickListener?.onFilmClick(mFilm?.id)
            } else {
                mClickListener?.onShowClick(mFilm?.id)
            }
            return true
        }

    }

    fun setClickListener(itemClickListener: ItemClickListener?) {
        mClickListener = itemClickListener
    }

    interface ItemClickListener {
        fun onFilmClick(id: String?)
        fun onShowClick(id: String?)
    }
}